File: client/src/main/java/net/treset/minecraftlauncher/update/UpdateService.java
package net.treset.minecraftlauncher.update;

import javafx.util.Pair;
import net.treset.minecraftlauncher.LauncherApplication;
import net.treset.minecraftlauncher.util.HttpService;
import net.treset.minecraftlauncher.util.HttpUtil;

import java.io.IOException;

public class UpdateService extends HttpService {
    public UpdateService() {
        super(LauncherApplication.config.UPDATE_URL);
    }

    public Update update() throws IOException {
        Pair<HttpUtil.HttpStatusCode, byte[]> result = get("update/" + LauncherApplication.stringLocalizer.get("launcher.version"));
        String response = new String(result.getValue());
        try {
            return Update.fromJson(response);
        } catch(Exception e) {
            throw new IOException("Failed to parse the response from the server.\nError: " + e);
        }
    }

    public byte[] file(String version, String file) throws IOException {
        Pair<HttpUtil.HttpStatusCode, byte[]> result = get("file/" + version + "/" + file);
        return result.getValue();
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/update/UpdateService.kt
package net.treset.treelauncher.backend.update

import net.treset.treelauncher.backend.config.appConfig
import net.treset.treelauncher.backend.util.HttpService
import net.treset.treelauncher.localization.language
import net.treset.treelauncher.localization.strings
import java.io.IOException

class UpdateService : HttpService(appConfig().updateUrl) {
    @Throws(IOException::class)
    fun update(): Update {
        val result: Pair<HttpStatusCode, ByteArray> = get(
            "update",
            strings().launcher.version(),
            language().appLanguage.locale
        )
        val response = String(result.second)
        return try {
            Update.fromJson(response)
        } catch (e: Exception) {
            throw IOException("Failed to parse the response from the server.\nError: $e")
        }
    }

    @Throws(IOException::class)
    fun file(version: String, file: String): ByteArray {
        val result: Pair<HttpStatusCode, ByteArray> = get("file", version, file)
        return result.second
    }
}
